File: clases/Cap03_Seleccion_de_Caracteristicas/python/arrows/p3.py
'''
Enunciado del ejercicio:
Diseñar un clasificador automático de flechas, que puede ser usado para
identificación de señales de tránsito. Los datos de training y testing 
están en los directorios arrows_training and arrows_testing respectivamente. 
Las imágenes se encuentran imágenes en formato PNG con tres clases de flechas:
1) unidireccional-curva     2) bidireccional     3) unidireccional-recta.

Usando características geometricas (momentos de Hu, momentos de Flusser y 
descriptores de Forurier) y la técnica de selección de características SFS, 
diseñe un clasificador KNN (con K=1) que clasifique correctamente 
las flechas de la carpeta arrows_testing.

(c) Solución proporcionada por: 
Germán Leandro Contreras Sagredo, Daniel Vives
Santiago de Chile, 04 de Abril de 2019
Universidad Católica de Chile
'''

# Librerías a utilizar

import numpy as np                                     # Manejo arreglos.
from pybalu.io import imread                           # Lectura imágenes.
from sklearn.neighbors import KNeighborsClassifier     # Modelo KNN
from pybalu.feature_extraction import (hugeo_features, # Extracción features.
flusser_features,fourier_features)
from pybalu.feature_selection import sfs               # Selección features.
from pybalu.performance_eval import performance        # Cómputo accuracy.

# Función para añadir las features al dataset.
def add_feats(data,path,im_class):
    # Se abre el archivo y se convierte en un numpy array.
    im = imread(path)
    # Se binariza.
    im_bin = (im > 140).astype(int)
    # Se extraen todas las características.
    hu_feat = hugeo_features(im_bin)
    flusser_feat = flusser_features(im_bin)
    fourier_feat = fourier_features(im_bin)
    # Se crea la fila y se añade la columna de la clase.
    row = np.concatenate((hu_feat,flusser_feat,fourier_feat,
                          np.array([im_class])))
    # Se añade la fila en el conjunto.
    if data.shape[0] == 0:
        return row
    return np.vstack([data,row])

# Arreglo de los datos
arrows_train = np.array([])
arrows_test = np.array([])

# Flechas curvas unidireccionales.
for i in range(1,13):
    path_train,path_test = '',''
    if i < 10:
        path_train = "arrows_training/arrow01_0{}.png".format(i)
        path_test = "arrows_testing/arrow01_0{}.png".format(i)
    else:
        path_train = "arrows_training/arrow01_{}.png".format(i)
        path_test = "arrows_testing/arrow01_{}.png".format(i)
    # Se añade la fila con la clase 1.
    arrows_train = add_feats(arrows_train,path_train,1)
    if i < 11:
        arrows_test = add_feats(arrows_test,path_test,1)

# Flechas bidireccionales.
for i in range(1,13):
    path_train,path_test = '',''
    if i < 10:
        path_train = "arrows_training/arrow02_0{}.png".format(i)
        path_test = "arrows_testing/arrow02_0{}.png".format(i)
    else:
        path_train = "arrows_training/arrow02_{}.png".format(i)
        path_test = "arrows_testing/arrow02_{}.png".format(i)
    # Se añade la fila con la clase 2.
    arrows_train = add_feats(arrows_train,path_train,2)
    if i < 11:
        arrows_test = add_feats(arrows_test,path_test,2)

# Flechas unidireccionales.
for i in range(1,13):
    path_train,path_test = '',''
    if i < 10:
        path_train = "arrows_training/arrow03_0{}.png".format(i)
        path_test = "arrows_testing/arrow03_0{}.png".format(i)
    else:
        path_train = "arrows_training/arrow03_{}.png".format(i)
        path_test = "arrows_testing/arrow03_{}.png".format(i)
    # Se añade la fila con la clase 3.
    arrows_train = add_feats(arrows_train,path_train,3)
    if i < 11:
        arrows_test = add_feats(arrows_test,path_test,3)


# Cantidad de features a utilizar.
N_FEATURES = 5

# Obtenemos X,Y de las features calculadas, tanto de entrenamiento como de test.
Xtrain,Ytrain = arrows_train[:, :-1],arrows_train[:, -1].astype(int)
Xtest,Ytest = arrows_test[:, :-1],arrows_test[:, -1].astype(int)


# Selección de features
selected_feats = sfs(Xtrain,Ytrain,n_features=N_FEATURES,method="fisher",
                     show=False)

# Se definen Xtrain_new,Xtest_new con las features seleccionadas.
Xtrain_new, Xtest_new = Xtrain[:,selected_feats],Xtest[:,selected_feats]

# Se define el modelo y se entrena.
knn = KNeighborsClassifier(n_neighbors=1)
knn.fit(Xtrain_new,Ytrain)

# Se obtiene la predicción
Ypred_new = knn.predict(Xtest_new)

# Computamos el accuracy.
result = performance(Ypred_new, Ytest)
print(result)

File: clases/Cap03_Seleccion_de_Caracteristicas/matlab/arrows/p3.m
% Enunciado del ejercicio:
% Disenar un clasificador automatico de flechas, que puede ser usado para
% identificacion de senales de tr?nsito. Los datos de training y testing 
% estan en los directorios arrows_training and arrows_testing respectivamente. 
% Las imagenes se encuentran im?genes en formato PNG con tres clases de flechas:
% 1) unidireccional-curva     2) bidireccional     3) unidireccional-recta.
% 
% Usando caracteristicas geometricas (momentos de Hu, momentos de Flusser y 
% descriptores de Forurier) y la tecnica de seleccion de caracter?sticas SFS, 
% disene un clasificador KNN (con K=1) que clasifique correctamente 
% las flechas de la carpeta arrows_testing.
% 
% (c) D. Mery
% Santiago de Chile, 05 de Abril de 2019
% Universidad Cat?lica de Chile


clt

% Features defintion
% Hu moments
b(1).name = 'hugeo';
b(1).options.show    = 1;
b(1).options.type    = 1;    % geometric

% Flusster moments
b(2).name = 'flusser';
b(2).options.show    = 1;
b(2).options.type    = 1;    % geometric

% Fourier descriptors
b(3).name = 'fourierdes';
b(3).options.show        = 1;
b(3).options.Nfourierdes = 16;
b(3).options.type        = 1;  % geometric
options.b = b;

nf = sum([7 4 b(3).options.Nfourierdes]); %number of features

% Training images definition
c = [12 12 12]; % 12 samples per class
n = sum(c);

% Extraction features from training images
Xtrain = zeros(n,nf);
st = 'arrows_training';
d = dir([st '/*.png']);
for i=1:n
    I = imread([st '/' d(i).name]);
    R = I>140;
    Xtrain(i,:) = Bfx_geo(R,options);
end
% Normalization
% [Xtrain,a,b] = Bft_norm(Xtrain,1);

% Training Labels
Ytrain = Bds_labels(c);

% Testing images definition
c = [10 10 10]; % 10 samples per class
n = sum(c);

% Extraction features from testing images
Xtest = zeros(n,nf);
st = 'arrows_testing';
d = dir([st '/*.png']);
for i=1:n
    I = imread([st '/' d(i).name]);
    R = I>140;
    Xtest(i,:) = Bfx_geo(R,options);
end
% Normalization
% Xtest = Xtest.*(ones(n,1)*a) + ones(n,1)*b;

% Testing Labels
Ytest = Bds_labels(c);

% Feature selection
op.m    = 3;                           % 3 features will be selected
op.show = 0;                           % don't display results
op.b.name = 'fisher';                  % SFS with Fisher
s = Bfs_sfs(Xtrain,double(Ytrain),op); % index of selected features

% Training and testing
op.k = 1;
Ypred = Bcl_knn(Xtrain(:,s),Ytrain,Xtest(:,s),op);
Bev_performance(Ypred,Ytest)